File: apps/macos/Sources/DocFlow/Engine/EngineModels.swift
import Foundation

// Wire types of the docflow-engine JSON-RPC protocol (engine/src/rpc.rs and
// engine/src/models.rs). EngineClient converts between snake_case and
// camelCase keys.

struct DocumentInfo: Decodable, Identifiable, Hashable, Sendable {
    var id: String
    var title: String
    var originalFilename: String
    var sourceSize: Int64
    var sourceSha256: String
    var mimeType: String?
    var processingMode: String
    var translationTier: Int
    /// "Google 翻译（免费）" or "服务商 · 模型"; absent for older documents.
    var translatorLabel: String?
    var mineruModel: String
    var status: String
    var stage: String
    var progress: Int
    var failureReason: String?
    var queueAttempts: Int
    var pagesProcessed: Int?
    var pagesTotal: Int?
    var imageCount: Int
    var excerpt: String?
    var translated: Bool
    var pdfSize: Int64?
    var dualPdfSize: Int64?
    var createdAt: Date
    var updatedAt: Date
    var startedAt: Date?
    var completedAt: Date?
    // Only present in documents.get / create / rename / retry / cancel.
    var files: DocumentFiles?
    var suggestedNames: SuggestedNames?
    var running: Bool?

    var isNative: Bool { processingMode == "pdf2zh" }
    var isActive: Bool { status == "queued" || status == "processing" || status == "retrying" }
    var isCompleted: Bool { status == "completed" }
    var isStopped: Bool { status == "failed" || status == "cancelled" }
}

struct DocumentFiles: Decodable, Hashable, Sendable {
    var source: String?
    var archiveDir: String?
    var journalPdf: String?
    var monoPdf: String?
    var dualPdf: String?
    var markdown: String?
    var markdownOriginal: String?
    var markdownTranslated: String?
    var readerHtml: String?
}

struct SuggestedNames: Decodable, Hashable, Sendable {
    var journalPdf: String
    var monoPdf: String
    var dualPdf: String
    var markdown: String
    var bundle: String
    var source: String
}

struct DocumentList: Decodable, Sendable {
    var items: [DocumentInfo]
    var total: Int
    var counts: StatusCounts
}

struct StatusCounts: Decodable, Hashable, Sendable {
    var all = 0
    var active = 0
    var completed = 0
    var failed = 0
}

struct ProcessingEvent: Decodable, Identifiable, Hashable, Sendable {
    var id: Int64
    var documentId: String
    var stage: String
    var state: String
    var level: String
    var progress: Int
    var message: String
    var detail: String?
    var current: Int64?
    var total: Int64?
    var createdAt: Date

    var isAttention: Bool { level == "warning" || level == "error" }
}

struct EventList: Decodable, Sendable {
    var items: [ProcessingEvent]
    var total: Int
    var nextAfterId: Int64
    var hasMore: Bool
}

struct InitializeResult: Decodable, Sendable {
    var version: String
    var dataDir: String
    var resourcesDir: String
    var settings: SettingsInfo
}

struct SettingsInfo: Decodable, Sendable {
    var preferences: Preferences
    var translationRuntime: TranslationRuntime
    var translationRuntimeDefaults: TranslationRuntime
    var translationRuntimeLimits: RuntimeLimits
    var providers: [ProviderInfo]
    var providerPresets: [ProviderPreset]
    var mineru: SecretStatus
    var capabilities: Capabilities
    var dataDir: String
    var engineVersion: String

    func provider(_ id: String) -> ProviderInfo? {
        providers.first { $0.id == id }
    }

    /// Google, then every model of every usable provider.
    var translatorOptions: [TranslatorOption] {
        var options = [TranslatorOption.google]
        for provider in providers where provider.isUsable(fake: capabilities.fakeProviders) {
            for model in provider.models {
                options.append(TranslatorOption(
                    choice: .llm(providerID: provider.id, model: model.id),
                    label: "\(provider.name) · \(model.name ?? model.id)"
                ))
            }
        }
        return options
    }
}

struct Preferences: Codable, Equatable, Sendable {
    var mineruModel: String
    var defaultTranslator: TranslatorChoice
    var defaultMode: String
    var proxy: ProxySettings
    var workerConcurrency: Int
}

/// `{"kind": "google"}` or `{"kind": "llm", "provider_id": …, "model": …}`.
/// Absent optionals are left out when encoding, as the engine requires.
struct TranslatorChoice: Codable, Hashable, Sendable {
    var kind: String
    var providerId: String?
    var model: String?

    static let google = TranslatorChoice(kind: "google")

    static func llm(providerID: String, model: String) -> TranslatorChoice {
        TranslatorChoice(kind: "llm", providerId: providerID, model: model)
    }

    var isLLM: Bool { kind == "llm" }
}

struct TranslatorOption: Identifiable, Hashable, Sendable {
    var choice: TranslatorChoice
    var label: String

    var id: TranslatorChoice { choice }

    static let google = TranslatorOption(choice: .google, label: "Google 翻译（免费）")
}

/// `{"mode": "system" | "direct"}` or `{"mode": "custom", "url": "…"}`.
struct ProxySettings: Codable, Equatable, Sendable {
    var mode: String
    var url: String?
}

struct TranslationRuntime: Codable, Equatable, Sendable {
    var google: GoogleRuntime
    var llm: LLMRuntime
    var perDocumentConcurrency: Int
    var systemPrompt: String
}

struct GoogleRuntime: Codable, Equatable, Sendable {
    var concurrency: Int
    var chunkChars: Int
}

struct LLMRuntime: Codable, Equatable, Sendable {
    var chunkChars: Int
    var maxSegmentsPerRequest: Int
    var maxRequestChars: Int
    var maxOutputTokens: Int
}

struct RuntimeLimits: Decodable, Sendable {
    var minChunkChars: Int
    var googleConcurrencyMax: Int
    var googleChunkCharsMax: Int
    var llmChunkCharsMax: Int
    var llmSegmentsPerRequestMax: Int
    var llmRequestCharsMin: Int
    var llmRequestCharsMax: Int
    var llmOutputTokensMax: Int
    var perDocumentConcurrencyMax: Int
    var providerConcurrencyMax: Int
    var systemPromptMaxChars: Int
}

struct ModelConfig: Codable, Hashable, Sendable {
    var id: String
    var name: String?
}

/// A large-model provider as the engine reports it. Its extra request
/// fields arrive as JSON text: the snake_case key strategies of
/// EngineCoding would rename keys inside an object (`generationConfig`).
struct ProviderInfo: Decodable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String
    var type: String
    var baseUrl: String
    var enabled: Bool
    var models: [ModelConfig]
    var concurrency: Int
    var preset: String?
    var extraBodyJson: String
    var keyConfigured: Bool
    var keyMasked: String?
    var keyOptional: Bool
    var keyUrl: String?

    func isUsable(fake: Bool) -> Bool {
        enabled && !models.isEmpty && (fake || keyConfigured || keyOptional)
    }

    var config: ProviderConfig {
        ProviderConfig(id: id, name: name, type: type, baseUrl: baseUrl, enabled: enabled, models: models, concurrency: concurrency, preset: preset)
    }
}

/// A provider as `providers.save` takes it (extra fields travel separately).
struct ProviderConfig: Encodable, Hashable, Sendable {
    var id: String
    var name: String
    var type: String
    var baseUrl: String
    var enabled: Bool
    var models: [ModelConfig]
    var concurrency: Int
    var preset: String?
}

struct ProviderPreset: Decodable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String
    var type: String
    var baseUrl: String
    var keyUrl: String
    var keyOptional: Bool
}

struct RemoteModel: Decodable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String?
    var owner: String?
    var contextLength: Int?
}

struct RemoteModels: Decodable, Sendable {
    var models: [RemoteModel]
}

struct CheckResult: Decodable, Sendable {
    var ok: Bool
    var latencyMs: Int
    var reply: String
}

struct SecretStatus: Decodable, Sendable {
    var configured: Bool
    var masked: String?
}

struct Capabilities: Decodable, Sendable {
    var mineruReady: Bool
    var pdf2zhReady: Bool
    var pdf2zhIssue: String?
    var googleReady: Bool
    var llmReady: Bool
    var fakeProviders: Bool
    var maxUploadMb: Int
    var mineruExtensions: [String]
    var pdf2zhExtensions: [String]
}

struct ExportResult: Decodable, Sendable {
    var path: String
    var bytes: Int64
}

/// A result whose content is not needed (`{}`, `{"ok": true}`, …).
struct Empty: Decodable, Sendable {}

// MARK: - Request parameters

struct IDParams: Encodable, Sendable {
    var id: String
}

struct ListParams: Encodable, Sendable {
    var filter: String?
    var query: String?
    var limit: Int?
}

struct EventsParams: Encodable, Sendable {
    var id: String
    var afterId: Int64
    var limit: Int
}

struct CreateParams: Encodable, Sendable {
    var path: String
    var title: String?
    var mode: String
    var translator: TranslatorChoice
}

struct RenameParams: Encodable, Sendable {
    var id: String
    var title: String
}

struct ExportParams: Encodable, Sendable {
    var id: String
    var destination: String
}

/// `mineru` or `provider:<id>`; `value: null` removes the key.
struct SecretParams: Encodable, Sendable {
    var name: String
    var value: String?

    // `value: null` must be written, not left out.
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(name, forKey: .name)
        try container.encode(value, forKey: .value)
    }

    private enum CodingKeys: String, CodingKey {
        case name, value
    }
}

struct InitializeParams: Encodable, Sendable {
    var secrets: [String: String]
}

struct SettingsUpdateParams: Encodable, Sendable {
    var preferences: Preferences?
    var translationRuntime: TranslationRuntime?
}

struct ProviderSaveParams: Encodable, Sendable {
    var provider: ProviderConfig
    /// Blank clears the extra request fields.
    var extraBodyJson: String
}

/// A provider endpoint for `providers.models` / `providers.check`; without
/// `apiKey` the engine uses the key stored for `providerId`.
struct EndpointParams: Encodable, Sendable {
    var providerId: String?
    var name: String
    var type: String
    var baseUrl: String
    var apiKey: String?
    var model: String?
    var extraBodyJson: String?
}

struct NoParams: Encodable, Sendable {}

File: apps/windows/DocFlow/Services/EngineModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace DocFlow.Services;

// Wire types of the docflow-engine JSON-RPC protocol (engine/src/rpc.rs).
// Property names map to snake_case through the client's naming policy.

public sealed class DocumentInfo
{
    public string Id { get; set; } = "";
    public string Title { get; set; } = "";
    public string OriginalFilename { get; set; } = "";
    public long SourceSize { get; set; }
    public string SourceSha256 { get; set; } = "";
    public string? MimeType { get; set; }
    public string ProcessingMode { get; set; } = "mineru";
    public int TranslationTier { get; set; }
    /// <summary>"服务商 · 模型" (3.0.0 documents may say "Google 翻译（免费）"); absent for older documents.</summary>
    public string? TranslatorLabel { get; set; }
    public string MineruModel { get; set; } = "";
    public string Status { get; set; } = "";
    public string Stage { get; set; } = "";
    public int Progress { get; set; }
    public string? FailureReason { get; set; }
    public int QueueAttempts { get; set; }
    public int? PagesProcessed { get; set; }
    public int? PagesTotal { get; set; }
    public int ImageCount { get; set; }
    public string? Excerpt { get; set; }
    public bool Translated { get; set; }
    public long? PdfSize { get; set; }
    public long? DualPdfSize { get; set; }
    public DateTimeOffset CreatedAt { get; set; }
    public DateTimeOffset UpdatedAt { get; set; }
    public DateTimeOffset? StartedAt { get; set; }
    public DateTimeOffset? CompletedAt { get; set; }
    public DocumentFiles? Files { get; set; }
    public SuggestedNames? SuggestedNames { get; set; }
    public bool Running { get; set; }

    [JsonIgnore]
    public bool IsNative => ProcessingMode == "pdf2zh";

    [JsonIgnore]
    public bool IsActive => Status is "queued" or "processing" or "retrying";
}

public sealed class DocumentFiles
{
    public string? Source { get; set; }
    public string? ArchiveDir { get; set; }
    public string? JournalPdf { get; set; }
    public string? MonoPdf { get; set; }
    public string? DualPdf { get; set; }
    public string? Markdown { get; set; }
    public string? MarkdownOriginal { get; set; }
    public string? MarkdownTranslated { get; set; }
    public string? ReaderHtml { get; set; }
}

public sealed class SuggestedNames
{
    public string JournalPdf { get; set; } = "";
    public string MonoPdf { get; set; } = "";
    public string DualPdf { get; set; } = "";
    public string Markdown { get; set; } = "";
    public string Bundle { get; set; } = "";
    public string Source { get; set; } = "";
}

public sealed class DocumentList
{
    public List<DocumentInfo> Items { get; set; } = [];
    public long Total { get; set; }
    public StatusCounts Counts { get; set; } = new();
}

public sealed class StatusCounts
{
    public long All { get; set; }
    public long Active { get; set; }
    public long Completed { get; set; }
    public long Failed { get; set; }
}

public sealed class ProcessingEvent
{
    public long Id { get; set; }
    public string DocumentId { get; set; } = "";
    public string Stage { get; set; } = "";
    public string State { get; set; } = "";
    public string Level { get; set; } = "info";
    public int Progress { get; set; }
    public string Message { get; set; } = "";
    public string? Detail { get; set; }
    public long? Current { get; set; }
    public long? Total { get; set; }
    public DateTimeOffset CreatedAt { get; set; }
}

public sealed class EventList
{
    public List<ProcessingEvent> Items { get; set; } = [];
    public long Total { get; set; }
    public long NextAfterId { get; set; }
    public bool HasMore { get; set; }
}

public sealed class InitializeResult
{
    public int Protocol { get; set; }
    public string Version { get; set; } = "";
    public string DataDir { get; set; } = "";
    public string ResourcesDir { get; set; } = "";
    public SettingsInfo Settings { get; set; } = new();
}

public sealed class SettingsInfo
{
    public Preferences Preferences { get; set; } = new();
    public TranslationRuntime TranslationRuntime { get; set; } = new();
    public TranslationRuntime TranslationRuntimeDefaults { get; set; } = new();
    public RuntimeLimits TranslationRuntimeLimits { get; set; } = new();
    public List<ProviderInfo> Providers { get; set; } = [];
    public List<ProviderPreset> ProviderPresets { get; set; } = [];
    public SecretStatus Mineru { get; set; } = new();
    public Capabilities Capabilities { get; set; } = new();
    public string DataDir { get; set; } = "";
    public string EngineVersion { get; set; } = "";

    public ProviderInfo? Provider(string id) => Providers.FirstOrDefault(provider => provider.Id == id);

    /// <summary>Every model a document can be translated with right now.</summary>
    public List<TranslatorOption> TranslatorOptions()
    {
        var options = new List<TranslatorOption>();
        foreach (var provider in Providers.Where(provider => provider.Usable(Capabilities.FakeProviders)))
        {
            options.AddRange(provider.Models.Select(model => new TranslatorOption(
                TranslatorChoice.Llm(provider.Id, model.Id),
                $"{provider.Name} · {model.Name ?? model.Id}")));
        }
        return options;
    }
}

public sealed class Preferences
{
    public string MineruModel { get; set; } = "vlm";

    /// <summary>Null until the user picks a model; "新建翻译" then offers the first one.</summary>
    public TranslatorChoice? DefaultTranslator { get; set; }

    public string DefaultMode { get; set; } = "pdf2zh";
    public ProxySettings Proxy { get; set; } = new();
    public int WorkerConcurrency { get; set; } = 2;
}

/// <summary><c>{"kind":"llm","provider_id":…,"model":…}</c>: a model of a provider.</summary>
public sealed class TranslatorChoice
{
    public string Kind { get; set; } = "llm";

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? ProviderId { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Model { get; set; }

    public static TranslatorChoice Llm(string providerId, string model) =>
        new() { Kind = "llm", ProviderId = providerId, Model = model };

    [JsonIgnore]
    public string Key => $"{Kind}:{ProviderId}:{Model}";
}

public sealed record TranslatorOption(TranslatorChoice Choice, string Label)
{
    public override string ToString() => Label;
}

public sealed class ProxySettings
{
    public string Mode { get; set; } = "system";

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Url { get; set; }
}

public sealed class TranslationRuntime
{
    public LlmRuntime Llm { get; set; } = new();
    public int PerDocumentConcurrency { get; set; }
    public string SystemPrompt { get; set; } = "";
}

public sealed class LlmRuntime
{
    public int ChunkChars { get; set; }
    public int MaxSegmentsPerRequest { get; set; }
    public int MaxRequestChars { get; set; }
    public long MaxOutputTokens { get; set; }
}

public sealed class RuntimeLimits
{
    public int MinChunkChars { get; set; }
    public int LlmChunkCharsMax { get; set; }
    public int LlmSegmentsPerRequestMax { get; set; }
    public int LlmRequestCharsMin { get; set; }
    public int LlmRequestCharsMax { get; set; }
    public double LlmOutputTokensMax { get; set; }
    public int PerDocumentConcurrencyMax { get; set; }
    public int ProviderConcurrencyMax { get; set; }
    public int SystemPromptMaxChars { get; set; }
}

public sealed class ModelConfig
{
    public string Id { get; set; } = "";

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Name { get; set; }
}

/// <summary>A provider as <c>providers.save</c> accepts it.</summary>
public class ProviderConfig
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string Type { get; set; } = "openai";
    public string BaseUrl { get; set; } = "";
    public bool Enabled { get; set; } = true;
    public List<ModelConfig> Models { get; set; } = [];
    public int Concurrency { get; set; } = 100;

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Preset { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public Dictionary<string, JsonElement>? ExtraBody { get; set; }
}

/// <summary>A provider as <c>settings.get</c> reports it, with its key status.</summary>
public sealed class ProviderInfo : ProviderConfig
{
    public bool KeyConfigured { get; set; }
    public string? KeyMasked { get; set; }
    public bool KeyOptional { get; set; }
    public string? KeyUrl { get; set; }

    public bool Usable(bool fake) => Enabled && Models.Count > 0 && (fake || KeyConfigured || KeyOptional);

    public ProviderConfig ToConfig() => new()
    {
        Id = Id,
        Name = Name,
        Type = Type,
        BaseUrl = BaseUrl,
        Enabled = Enabled,
        Models = Models.Select(model => new ModelConfig { Id = model.Id, Name = model.Name }).ToList(),
        Concurrency = Concurrency,
        Preset = Preset,
        ExtraBody = ExtraBody,
    };
}

public sealed class ProviderPreset
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string Type { get; set; } = "openai";
    public string BaseUrl { get; set; } = "";
    public string KeyUrl { get; set; } = "";
    public bool KeyOptional { get; set; }
}

public sealed class RemoteModel
{
    public string Id { get; set; } = "";
    public string? Name { get; set; }
    public string? Owner { get; set; }
    public long? ContextLength { get; set; }
}

public sealed class RemoteModels
{
    public List<RemoteModel> Models { get; set; } = [];
}

public sealed class CheckResult
{
    public bool Ok { get; set; }
    public long LatencyMs { get; set; }
    public string Reply { get; set; } = "";
}

public sealed class SecretStatus
{
    public bool Configured { get; set; }
    public string? Masked { get; set; }
}

public sealed class Capabilities
{
    public bool MineruReady { get; set; }
    public bool Pdf2zhReady { get; set; }
    public string? Pdf2zhIssue { get; set; }
    public bool LlmReady { get; set; }
    public bool FakeProviders { get; set; }
    public long MaxUploadMb { get; set; }
    public List<string> MineruExtensions { get; set; } = [];
    public List<string> Pdf2zhExtensions { get; set; } = [];
}
public sealed class ExportResult
{
    public string Path { get; set; } = "";
    public long Bytes { get; set; }
}
